### ISSUE TITLE
Asset compilation fails -- 'error:03000086:digital envelope routines::initialization error'

### ISSUE THREAD

**Issue by jamincollins** [mastodon/mastodon#20705]
### Steps to reproduce the problem

1. Install mastodon on Arch Linux
2. run  `RAILS_ENV=production bundle exec rails mastodon:setup`
...

Part of the above process appears to be compiling assets.

### Expected behaviour

Setup should work

### Actual behaviour

Aborts with the above error and a stack trace

### Detailed description

```
...
Compile the assets now? Yes
Running `RAILS_ENV=production rails assets:precompile` ...


yarn install v1.22.19
[1/6] Validating package.json...
[2/6] Resolving packages...
[3/6] Fetching packages...
[4/6] Linking dependencies...
warning Workspaces can only be enabled in private projects.
[5/6] Building fresh packages...
[6/6] Cleaning modules...
Done in 81.48s.
I, [2022-11-14T21:10:50.256575 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/doorkeeper/admin/application-a644908e7bab54fb749be0f59fb64a7480bbf9c4c2b79d4a65791cb7ab4d8730.css
I, [2022-11-14T21:10:50.257497 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/doorkeeper/admin/application-a644908e7bab54fb749be0f59fb64a7480bbf9c4c2b79d4a65791cb7ab4d8730.css.gz
I, [2022-11-14T21:10:50.274947 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/doorkeeper/application-c93dac2ad9d65e3393e0e2c958481e86ef7a5e5b0f6ce406842a7b99b25a4850.css
I, [2022-11-14T21:10:50.275461 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/doorkeeper/application-c93dac2ad9d65e3393e0e2c958481e86ef7a5e5b0f6ce406842a7b99b25a4850.css.gz
I, [2022-11-14T21:10:50.324448 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/pghero/favicon-db10337a56c45eb43c22ff5019546b520fa22c7281d4d385f235cbca67ed26bb.png
I, [2022-11-14T21:10:51.289770 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/pghero/application-a60bf0a452ed064fef3594cf52a4c998712da7c76150f890f4eaa644f59671e4.js
I, [2022-11-14T21:10:51.290177 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/pghero/application-a60bf0a452ed064fef3594cf52a4c998712da7c76150f890f4eaa644f59671e4.js.gz
I, [2022-11-14T21:10:51.345289 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/pghero/application-c31338f656687c1d733bb0f48d40acd076e24060f3dcff83b34870e4ccc2789d.css
I, [2022-11-14T21:10:51.345938 #6691]  INFO -- : Writing /var/lib/mastodon/public/assets/pghero/application-c31338f656687c1d733bb0f48d40acd076e24060f3dcff83b34870e4ccc2789d.css.gz
Warning: you are using an unstable release of Node.js (v19.1.0). If you encounter issues with Node.js, consider switching to an Active LTS release. More info: https://docs.npmjs.com/try-the-latest-stable-version-of-node
Compiling...
Compilation failed:
node:internal/crypto/hash:71
  this[kHandle] = new _Hash(algorithm, xofLen);
                  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:71:19)
    at Object.createHash (node:crypto:140:10)
    at module.exports (/var/lib/mastodon/node_modules/webpack/lib/util/createHash.js:135:53)
    at ConcatenatedModule._createIdentifier (/var/lib/mastodon/node_modules/webpack/lib/optimize/ConcatenatedModule.js:563:16)
    at new ConcatenatedModule (/var/lib/mastodon/node_modules/webpack/lib/optimize/ConcatenatedModule.js:445:27)
    at /var/lib/mastodon/node_modules/webpack/lib/optimize/ModuleConcatenationPlugin.js:250:26
    at SyncBailHook.eval [as call] (eval at create (/var/lib/mastodon/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at SyncBailHook.lazyCompileHook (/var/lib/mastodon/node_modules/tapable/lib/Hook.js:154:20)
    at /var/lib/mastodon/node_modules/webpack/lib/Compilation.js:1351:37
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/var/lib/mastodon/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (/var/lib/mastodon/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.seal (/var/lib/mastodon/node_modules/webpack/lib/Compilation.js:1342:27)
    at /var/lib/mastodon/node_modules/webpack/lib/Compiler.js:675:18
    at /var/lib/mastodon/node_modules/webpack/lib/Compilation.js:1261:4
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/var/lib/mastodon/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:24:1)
    at AsyncSeriesHook.lazyCompileHook (/var/lib/mastodon/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.finish (/var/lib/mastodon/node_modules/webpack/lib/Compilation.js:1253:28)
    at /var/lib/mastodon/node_modules/webpack/lib/Compiler.js:672:17
    at eval (eval at create (/var/lib/mastodon/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
    at /var/lib/mastodon/node_modules/webpack/lib/Compilation.js:1185:12
    at /var/lib/mastodon/node_modules/webpack/lib/Compilation.js:1097:9
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}
```

### Specifications

Mastodon 4.0.0
Ruby 3.0.4p208
Node.s 19.1.0

It seems that this might be avoided with:

`NODE_OPTIONS=--openssl-legacy-provider`

https://github.com/webpack/webpack/issues/14532

**Comment by ineffyble** [comment]
Technically Node 16 is the supported version at moment. There's an open PR to bump to Node 18 which is hitting this error: #20586.

**Comment by ClearlyClaire** [comment]
I think this has to do with OpenSSL 3. Unfortunately, we're still using an old version of webpack which means we have to use algorithms which have since been dropped from OpenSSL 3 if I remember correctly.

**Comment by bholst** [comment]
@ClearlyClaire Could you add that information to the requirements section of the README?
https://github.com/mastodon/mastodon#requirements

**Comment by vmstan** [comment]
@jamincollins can you confirm that this issue can be closed? Mastodon 4.2 now has support for Node 18+

**Comment by mjankowski** [comment]
Closing on assumption that newer versions (of almost everything?! - mastodon, ruby, node) may have resolved this -- please reopen with steps to reproduce and/or logs if not.
